WireDomain/Sources/WireDomain/Synchronization/PullPendingUpdateEventsV2Sync.swift

```diff
@@ -0,0 +1,201 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+import WireCoreCrypto
+import WireDataModel
+import WireLogging
+import WireNetwork
+
+public struct PullPendingUpdateEventsSyncV2: PullPendingUpdateEventsSyncV2Protocol {
+
+    enum Failure: Error {
+        case acknowledgeFailed
+    }
+
+    private let selfClientID: String
+    private let pushChannelAPI: any PushChannelV2API
+    private let updateEventsStore: any UpdateEventsLocalStoreProtocol
+    private let journal: Journal
+    private let decryptor: any UpdateEventDecryptorProtocol
+    private let coreCryptoProvider: any CoreCryptoProviderProtocol
+    private let jsonEncoder = JSONEncoder()
+    private let logger = WireLogger.sync
+
+    let stream: AsyncStream<[UpdateEvent]>
+    private let continuation: AsyncStream<[UpdateEvent]>.Continuation
+
+    public init(
+        selfClientID: String,
+        pushChannelAPI: any PushChannelV2API,
+        updateEventsStore: any UpdateEventsLocalStoreProtocol,
+        journal: Journal,
+        decryptor: any UpdateEventDecryptorProtocol,
+        coreCryptoProvider: any CoreCryptoProviderProtocol
+    ) {
+        self.selfClientID = selfClientID
+        self.pushChannelAPI = pushChannelAPI
+        self.updateEventsStore = updateEventsStore
+        self.journal = journal
+        self.decryptor = decryptor
+        self.coreCryptoProvider = coreCryptoProvider
+
+        let (finalStream, continuation) = AsyncStream<[UpdateEvent]>.makeStream()
+        self.stream = finalStream
+        self.continuation = continuation
+    }
+
+    private var logAttributes: WireLogging.LogAttributes {
+        .syncAttributes(initialSync: true) + .newNSE
+    }
+
+    public func pull() async throws {
+        let pushChannel = try await pushChannelAPI.createPushChannel(clientID: selfClientID)
+
+        let liveEventStream = try await pushChannel.open()
+
+        logger.debug("handling live event stream", attributes: logAttributes)
+        do {
+            streamLoop: for try await element in liveEventStream {
+                try Task.checkCancellation()
+
+                logger.debug(
+                    "received live element: \(element)",
+                    attributes: .syncAttributes(initialSync: false)
+                )
+                switch element {
+                case .upToDate:
+                    logger.debug("upToDate event", attributes: logAttributes)
+                    continuation.finish()
+                    break streamLoop
```

Comment:
here we're upToDate so we close the stream we produce so notifications are generated + break from the forloop to close the pushchannel so NSE can finish and terminate after showing the notifications